jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/MavenProjectProperties.java

```diff
@@ -197,36 +199,42 @@ static int getVersionFromString(String versionString) {
 
   @Override
   public JibContainerBuilder createContainerBuilder(
-      RegistryImage baseImage, AbsoluteUnixPath appRoot) throws IOException {
+      RegistryImage baseImage, AbsoluteUnixPath appRoot, ContainerizingMode containerizingMode)
+      throws IOException {
+    JavaContainerBuilder javaContainerBuilder =
+        JavaContainerBuilder.from(baseImage).setAppRoot(appRoot);
+
     try {
       if (isWarProject()) {
         Path explodedWarPath =
-            Paths.get(project.getBuild().getDirectory()).resolve(project.getBuild().getFinalName());
-        return JavaContainerBuilderHelper.fromExplodedWar(baseImage, explodedWarPath, appRoot);
+            Paths.get(project.getBuild().getDirectory(), project.getBuild().getFinalName());
+        return JavaContainerBuilderHelper.fromExplodedWar(javaContainerBuilder, explodedWarPath);
+      }
+
+      if (containerizingMode == ContainerizingMode.PACKAGED) {
+        String jarName = project.getBuild().getFinalName() + "." + project.getPackaging();
+        Path jar = Paths.get(project.getBuild().getDirectory(), jarName);
```

Comment:
I could not find any other way. I believe it is the Maven JAR plugin (in the usual case) that packages a JAR, so I think the Maven API doesn't really have such a method. Looking at https://maven.apache.org/plugins/maven-jar-plugin/jar-mojo.html, I think the JAR name is fixed unless `<classifier>` is used. That said, this code is incomplete in that if either `<classifier>` or `<outputDirectory>` is defined, this won't work. In that case, I think we should really parse the JAR plugin config, but I'm punting that for the moment. Basically the same goes with the WAR case.

For the fat-JAR case, unfortunately it will end up duplicating dependencies. I think we should improve this, for example, by identifying if it is a fat JAR with various heuristics and skipping creating a dep layer. If you can think of other potential issues or edge cases, we should discuss too. But basically, we should think about how we can improve this going forward.

And I don't really grasp what kind of problems we will see with shaded JARs. Do you something specific as an example?

Reply:
Fat-JAR: actually, I think the only way to run it without us putting a dependency layer is to do `java -jar fat.jar`. That is, we go into the more complicated runnable JAR business.

Reply:
[`MavenProject#getAttachedArtifacts()`](https://github.com/apache/maven/blob/master/maven-core/src/main/java/org/apache/maven/project/MavenProject.java#L927) will return the list of artifacts that have been created for this project.  The shade plugin allows [replacing the artifact via the `shadedArtifactAttached` property](https://maven.apache.org/plugins/maven-shade-plugin/shade-mojo.html#shadedArtifactAttached).  I don't know whether it's worth asking users for the classifier of the artifact to be included.

(Aside: I think Jib should be attaching the result of `buildTar`.)

That said, we still won't know whether the jar is a fat jar or a slim jar (i.e., needs dependencies).  Perhaps we should have a third containerizing-mode that indicates that dependencies are ignored.  `standalone`?

Reply:
Ah, that's interesting. I'll look into the method.

> (Aside: I think Jib should be attaching the result of `buildTar`.)

Can you elaborate on it? I don't think I get the context. What is the result and where do we attach it?

> That said, we still won't know whether the jar is a fat jar or a slim jar (i.e., needs dependencies). Perhaps we should have a third containerizing-mode that indicates that dependencies are ignored. `standalone`?

Yeah, we need to see if there are reasonable or limited ways to check if it is a fat jar, and depending on the outcome, we may need to add another mode for that. But I think just skipping dependencies won't just work with `java -cp ... main.Class`. In any case, I think this PR is good for the first iteration and reasonable complete in the non-fat case.

Reply:
> Can you elaborate on it? I don't think I get the context. What is the result and where do we attach it?

Never mind. I get it now.

Reply:
Unfortunately, `MavenProject#getAttachedArtifacts()` is returning an empty list.

Reply:
Ah, I think you're considering a jar that contains other jars and whose manifest has a `Class-Path` attribute.  Whereas I'm considering a jar that includes *the contents* of other jars.

+1 to addressing this in subsequent PRs.  We might be able to examine the jar and determine if it's a fat/uberjar.

Reply:
Ah, the maven-jar-plugin uses attach-artifact if there's a classifier involved, otherwise it sets the file on the project artifact.  Here's the relevant code:

https://github.com/apache/maven-jar-plugin/blob/80f58a84aacff6e671f5a601d62a3a3800b507dc/src/main/java/org/apache/maven/plugins/jar/AbstractJarMojo.java#L295-L307

Reply:
Ah, now I understand what is going on. There's still an issue. In my previous testing, `project.getArtifact()` was returning an `Artifact` instance, but `artifact.getFile()` was returning null. However, I realized if you run the "package" goal, `setFile()` is called and `artifact.getFile()` returns the actual path to the JAR. That is, if the JAR plugin is executed, we're good, but we cannot require the user always combine "package" with "jib:build". Still no good alternative to get the path to the generated JAR.

Reply:
For `containerizerMode=packaged` I think requiring a `package` before `jib:build` is acceptable 😄  I suppose we could automatically invoke `package` if the artifact is not present.  But an error seems more appropriate.

Reply:
Resolved. FTR, the issue was that `package` should always be combined with a Jib goal at the time of running `mvn` for `getArtifact()` to return non-null `File`. See https://github.com/GoogleContainerTools/jib/pull/1746#discussion_r290355384